File: sdks/py/src/jeffs_brain_memory/search/trigram.py
# SPDX-License-Identifier: Apache-2.0
"""Trigram index for the fuzzy fallback leg.

Ported from the Go SDK's ``sdks/go/search/trigram.go``. The index maps
boundary-padded 3-grams to the slugs that contain them. Jaccard
similarity over the trigram sets drives ranking; the default threshold
of ``0.3`` matches ``spec/ALGORITHMS.md`` (Trigram fallback details).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

__all__ = [
    "TrigramIndex",
    "TrigramHit",
    "jaccard",
    "trigrams",
    "slug_text",
    "TRIGRAM_JACCARD_THRESHOLD",
]

TRIGRAM_JACCARD_THRESHOLD: float = 0.3
"""Minimum Jaccard similarity a slug must clear to survive the fallback."""

_DEFAULT_FUZZY_TOPK: int = 10


@dataclass(frozen=True, slots=True)
class TrigramHit:
    """One fuzzy match.

    Similarity is Jaccard over the padded trigram sets, in ``[0, 1]``.
    """

    path: str
    score: float


def trigrams(text: str) -> set[str]:
    """Return the set of boundary-padded 3-grams for ``text``.

    Text is lowercased and non-alphanumeric characters are replaced
    with spaces. Each whitespace-separated word is padded with a ``$``
    boundary marker at both ends so short-slug matches retain their
    word-boundary signal. Words shorter than three characters after
    padding are dropped.
    """
    out: set[str] = set()
    if not text:
        return out
    buf: list[str] = []
    for ch in text.lower():
        if ch.isalnum():
            buf.append(ch)
        else:
            buf.append(" ")
    cleaned = "".join(buf)
    for word in cleaned.split():
        padded = f"${word}$"
        if len(padded) < 3:
            continue
        for i in range(len(padded) - 2):
            out.add(padded[i : i + 3])
    return out


def jaccard(a: set[str], b: set[str]) -> float:
    """Compute ``|A ∩ B| / |A ∪ B|`` for two trigram sets."""
    if not a or not b:
        return 0.0
    if len(b) < len(a):
        a, b = b, a
    intersection = sum(1 for g in a if g in b)
    union = len(a) + len(b) - intersection
    if union == 0:
        return 0.0
    return intersection / union


def slug_text(path: str) -> str:
    """Return the whitespace-separated trigram source for ``path``.

    Mirrors Go's ``slugText``: lowercase, keep the last slash-separated
    segment, strip a trailing ``.md``, and collapse non-alphanumerics
    to spaces. ``clients/oude-reimer.md`` becomes ``oude reimer``.
    """
    s = path.lower()
    if "/" in s:
        s = s.rsplit("/", 1)[1]
    if s.endswith(".md"):
        s = s[:-3]
    buf: list[str] = []
    for ch in s:
        buf.append(ch if ch.isalnum() else " ")
    return " ".join("".join(buf).split())


class TrigramIndex:
    """Lazy in-memory trigram index keyed by slug text.

    Thread-safety: build-once-then-read. Mutation after
    :meth:`__init__` is not supported.
    """

    __slots__ = ("_index", "_paths", "_path_grams")

    def __init__(self, paths: Iterable[str]) -> None:
        self._index: dict[str, list[str]] = {}
        self._paths: list[str] = []
        self._path_grams: dict[str, set[str]] = {}

        seen: set[str] = set()
        for raw in paths:
            if not raw or raw in seen:
                continue
            seen.add(raw)
            self._paths.append(raw)
            grams = trigrams(slug_text(raw))
            self._path_grams[raw] = grams
            for gram in grams:
                self._index.setdefault(gram, []).append(raw)

    @property
    def paths(self) -> list[str]:
        """Return a fresh copy of the indexed slug set."""
        return list(self._paths)

    def fuzzy_search(
        self,
        query: str,
        top_k: int = _DEFAULT_FUZZY_TOPK,
        *,
        threshold: float = TRIGRAM_JACCARD_THRESHOLD,
    ) -> list[TrigramHit]:
        """Return the top-k slugs ranked by Jaccard similarity.

        ``threshold`` defaults to :data:`TRIGRAM_JACCARD_THRESHOLD`.
        Slugs below the threshold are discarded. Ties break on path
        ascending so the order is deterministic across runs.
        """
        if not self._index:
            return []
        if top_k <= 0:
            top_k = _DEFAULT_FUZZY_TOPK

        query_grams = trigrams(query)
        if not query_grams:
            return []

        candidates: set[str] = set()
        for gram in query_grams:
            for path in self._index.get(gram, ()):
                candidates.add(path)
        if not candidates:
            return []

        hits: list[TrigramHit] = []
        for path in candidates:
            grams = self._path_grams.get(path)
            if not grams:
                continue
            sim = jaccard(query_grams, grams)
            if sim < threshold:
                continue
            hits.append(TrigramHit(path=path, score=sim))

        hits.sort(key=lambda h: (-h.score, h.path))
        if len(hits) > top_k:
            hits = hits[:top_k]
        return hits

File: go/search/trigram.go
// SPDX-License-Identifier: Apache-2.0

package search

import (
	"sort"
	"strings"
	"unicode"
)

// TrigramIndex is a lazy in-memory index mapping trigrams to article
// slugs. Built from the set of wiki paths the first time a fuzzy
// lookup is requested. Cheap because we only store slugs, not content.
type TrigramIndex struct {
	// index maps a trigram to the list of slugs containing it.
	index map[string][]string
	// paths is the full slug set the index was built from, kept for
	// result shaping and introspection.
	paths []string
	// pathTrigrams caches the trigram set for each slug so we compute
	// it once at build time and reuse on every query.
	pathTrigrams map[string]map[string]struct{}
}

// TrigramHit is a single fuzzy match: the slug that matched plus the
// Jaccard similarity against the query text.
type TrigramHit struct {
	Path       string
	Similarity float64 // 0-1, Jaccard over trigram sets
}

// defaultFuzzyThreshold is the minimum Jaccard similarity a slug must
// clear before FuzzySearch will return it. 0.3 keeps single-character
// typos (Dude to Oude) inside the net while rejecting unrelated slugs
// that share only one or two trigrams.
const defaultFuzzyThreshold = 0.3

// BuildTrigramIndex walks every supplied path and builds the trigram
// map. "bosch" produces {"$bo", "bos", "osc", "sch", "ch$"} where $ is
// a boundary marker that biases matches toward word start and end.
func BuildTrigramIndex(paths []string) *TrigramIndex {
	idx := &TrigramIndex{
		index:        make(map[string][]string),
		paths:        make([]string, 0, len(paths)),
		pathTrigrams: make(map[string]map[string]struct{}, len(paths)),
	}
	seenPath := make(map[string]bool, len(paths))
	for _, p := range paths {
		if p == "" || seenPath[p] {
			continue
		}
		seenPath[p] = true
		idx.paths = append(idx.paths, p)

		tri := trigrams(slugText(p))
		idx.pathTrigrams[p] = tri
		for gram := range tri {
			idx.index[gram] = append(idx.index[gram], p)
		}
	}
	return idx
}

// Paths returns the slug set the index was built from.
func (t *TrigramIndex) Paths() []string {
	if t == nil {
		return nil
	}
	out := make([]string, len(t.paths))
	copy(out, t.paths)
	return out
}

// FuzzySearch returns the top-k slugs ranked by trigram overlap with
// query, filtered to a minimum similarity threshold. Uses Jaccard
// similarity: |Q intersect D| / |Q union D|. The threshold defaults
// to 0.3; slugs scoring below it are discarded.
func (t *TrigramIndex) FuzzySearch(query string, k int) []TrigramHit {
	return t.fuzzySearchWithThreshold(query, k, defaultFuzzyThreshold)
}

// fuzzySearchWithThreshold is the underlying implementation so tests
// can exercise alternative thresholds without re-implementing the
// ranking logic.
func (t *TrigramIndex) fuzzySearchWithThreshold(query string, k int, threshold float64) []TrigramHit {
	if t == nil || len(t.index) == 0 {
		return nil
	}
	if k <= 0 {
		k = 10
	}

	queryGrams := trigrams(query)
	if len(queryGrams) == 0 {
		return nil
	}

	// Collect candidate slugs that share at least one trigram with the
	// query. We only score candidates so the full corpus is not walked
	// for every fuzzy lookup.
	candidates := make(map[string]struct{})
	for gram := range queryGrams {
		for _, p := range t.index[gram] {
			candidates[p] = struct{}{}
		}
	}
	if len(candidates) == 0 {
		return nil
	}

	hits := make([]TrigramHit, 0, len(candidates))
	for p := range candidates {
		pathGrams := t.pathTrigrams[p]
		if len(pathGrams) == 0 {
			continue
		}
		sim := jaccard(queryGrams, pathGrams)
		if sim < threshold {
			continue
		}
		hits = append(hits, TrigramHit{Path: p, Similarity: sim})
	}

	sort.SliceStable(hits, func(i, j int) bool {
		if hits[i].Similarity != hits[j].Similarity {
			return hits[i].Similarity > hits[j].Similarity
		}
		return hits[i].Path < hits[j].Path
	})

	if len(hits) > k {
		hits = hits[:k]
	}
	return hits
}

// jaccard computes |A intersect B| / |A union B| for two trigram
// sets.
func jaccard(a, b map[string]struct{}) float64 {
	if len(a) == 0 || len(b) == 0 {
		return 0
	}
	var intersection int
	smaller, larger := a, b
	if len(b) < len(a) {
		smaller, larger = b, a
	}
	for g := range smaller {
		if _, ok := larger[g]; ok {
			intersection++
		}
	}
	union := len(a) + len(b) - intersection
	if union == 0 {
		return 0
	}
	return float64(intersection) / float64(union)
}

// trigrams returns the set of boundary-padded 3-grams for text,
// lowercased, with non-alphanumeric characters replaced by spaces.
// Each whitespace-separated word is padded with a single boundary
// marker ($) at the start and end so matches inside short slugs keep
// their word-boundary signal.
func trigrams(text string) map[string]struct{} {
	out := make(map[string]struct{})
	if text == "" {
		return out
	}

	var b strings.Builder
	for _, r := range strings.ToLower(text) {
		switch {
		case unicode.IsLetter(r), unicode.IsDigit(r):
			b.WriteRune(r)
		default:
			b.WriteRune(' ')
		}
	}
	cleaned := b.String()

	for _, word := range strings.Fields(cleaned) {
		padded := "$" + word + "$"
		runes := []rune(padded)
		if len(runes) < 3 {
			continue
		}
		for i := 0; i+3 <= len(runes); i++ {
			out[string(runes[i:i+3])] = struct{}{}
		}
	}

	return out
}

// slugText converts a wiki-relative path to a whitespace-separated
// token string suitable for trigram generation. Only the filename
// stem is considered so short queries that match the slug itself are
// not drowned out by parent-directory noise:
// "clients/oude-reimer.md" becomes "oude reimer".
func slugText(p string) string {
	s := strings.ToLower(p)
	if i := strings.LastIndex(s, "/"); i >= 0 {
		s = s[i+1:]
	}
	s = strings.TrimSuffix(s, ".md")
	var b strings.Builder
	for _, r := range s {
		switch {
		case unicode.IsLetter(r), unicode.IsDigit(r):
			b.WriteRune(r)
		default:
			b.WriteRune(' ')
		}
	}
	return strings.Join(strings.Fields(b.String()), " ")
}
